File: EduTrack/Services/AssignmentService.cs
using EduTrack.Models;
using EduTrackDataAccess.Entities;
using EduTrackDataAccess.Repositories.Assignments;

namespace EduTrack.Services
{
    public class AssignmentService : IGenericService<AssignmentModel>
    {
        private readonly IAssignmentRepository _repository;
        public AssignmentService(IAssignmentRepository repository)
        {
            _repository = repository;
        }

        public async Task<AssignmentModel> Create(AssignmentModel model)
        {
            var assignments = new Assignment
            {
                Id = model.Id,
                SubjectId = model.SubjectId,
                GroupId = model.GroupId,
                TeacherId = model.TeacherId,
                Title = model.Title,
                Deadline = model.Deadline,

            };
            var createdAssignments = await _repository.CreateAssignment(assignments);
            var result = new AssignmentModel()
            {
                Id = createdAssignments.Id,
                SubjectId= createdAssignments.SubjectId,    
                GroupId= createdAssignments.GroupId,
                TeacherId= createdAssignments.TeacherId,
                Title = createdAssignments.Title,
                Deadline = createdAssignments.Deadline,
            };
            return result;
        }

        public async Task<bool> Delete(int id)
        {
            return await _repository.DeleteAssignment(id);
        }

        public async Task<AssignmentModel> Get(int id)
        {
            var assignments = await _repository.GetAssignment(id);
            var model = new AssignmentModel
            {
                Id = assignments.Id,
                SubjectId= assignments.SubjectId,
                GroupId= assignments.GroupId,
                TeacherId= assignments.TeacherId,
                Title = assignments.Title,
                Deadline = assignments.Deadline,
            };
            return model;

        }

        public async Task<IEnumerable<AssignmentModel>> GetAll()
        {
            var result = new List<AssignmentModel>();
            var assignments = await _repository.GetAllAssignments();
            foreach (var assignment in assignments)
            {
                var model = new AssignmentModel
                {
                    Id = assignment.Id,
                    SubjectId= assignment.SubjectId,
                    GroupId= assignment.GroupId,
                    TeacherId = assignment.TeacherId,
                    Title = assignment.Title,
                    Deadline = assignment.Deadline,

                };
                result.Add(model);
            }
            return result;
        }

        public async Task<AssignmentModel> Update(int id, AssignmentModel model)
        {
            var assignments = new Assignment
            {
                Id = model.Id,
                SubjectId = model.SubjectId,
                GroupId = model.GroupId,
                TeacherId = model.TeacherId,
                Title = model.Title,
                Deadline = model.Deadline,
            };
            var updadedAssignments = await _repository.UpdateAssignment(id, assignments);
            var result = new AssignmentModel
            {
                Id = updadedAssignments.Id,
                SubjectId= updadedAssignments.SubjectId,
                GroupId= updadedAssignments.GroupId,
                TeacherId= updadedAssignments.TeacherId,
                Title = updadedAssignments.Title,
                Deadline = updadedAssignments.Deadline,
            };
            return result;
        }
    }
}

File: EduTrack/ClientApp/src/app/features/cabinet/teacher/services/assignment.service.ts
import { Injectable, inject, signal } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap, map } from 'rxjs';
import {
  AssignmentDto,
  Assignment,
  CreateAssignmentDto,
  UpdateAssignmentDto,
  AssignmentStatus
} from '../models/assignment.model';

@Injectable({
  providedIn: 'root'
})
export class AssignmentService {
  private http = inject(HttpClient);

  // Signal-based state
  private assignmentsSignal = signal<Assignment[]>([]);
  assignments = this.assignmentsSignal.asReadonly();

  private loadingSignal = signal(false);
  loading = this.loadingSignal.asReadonly();

  getAll(): Observable<Assignment[]> {
    this.loadingSignal.set(true);
    return this.http.get<AssignmentDto[]>('/assignments').pipe(
      map(dtos => dtos.map(dto => this.mapToModel(dto))),
      tap(assignments => {
        this.assignmentsSignal.set(assignments);
        this.loadingSignal.set(false);
      })
    );
  }

  getById(id: number): Observable<Assignment> {
    return this.http.get<AssignmentDto>(`/assignments/${id}`).pipe(
      map(dto => this.mapToModel(dto))
    );
  }

  create(data: CreateAssignmentDto): Observable<Assignment> {
    return this.http.post<AssignmentDto>('/assignments', data).pipe(
      map(dto => this.mapToModel(dto)),
      tap(assignment => {
        this.assignmentsSignal.update(assignments => [...assignments, assignment]);
      })
    );
  }

  update(id: number, data: UpdateAssignmentDto): Observable<Assignment> {
    return this.http.put<AssignmentDto>(`/assignments/${id}`, data).pipe(
      map(dto => this.mapToModel(dto)),
      tap(updatedAssignment => {
        this.assignmentsSignal.update(assignments =>
          assignments.map(a => a.id === id ? updatedAssignment : a)
        );
      })
    );
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`/assignments/${id}`).pipe(
      tap(() => {
        this.assignmentsSignal.update(assignments =>
          assignments.filter(a => a.id !== id)
        );
      })
    );
  }

  private mapToModel(dto: AssignmentDto): Assignment {
    const deadline = new Date(dto.deadline);
    const now = new Date();
    
    let status: AssignmentStatus;
    if (deadline < now) {
      status = AssignmentStatus.Expired;
    } else if (deadline.getTime() - now.getTime() > 7 * 24 * 60 * 60 * 1000) {
      status = AssignmentStatus.Upcoming;
    } else {
      status = AssignmentStatus.Active;
    }

    return {
      id: dto.id,
      title: dto.title,
      deadline: deadline,
      subjectId: dto.subjectId,
      groupId: dto.groupId,
      teacherId: dto.teacherId,
      status
    };
  }
}
